libraries/AP_SerialLED/AP_SerialLED.h

```diff
@@ -0,0 +1,47 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  AP_SerialLED for controlling serial connected LEDs using WS2812B protocol
+ */
+#pragma once
+
+#include <AP_HAL/AP_HAL.h>
+
+class AP_SerialLED {
+public:
+    AP_SerialLED();
+
+    // set number of LEDs per pin
+    bool set_num_LEDs(uint8_t chan, uint8_t num_leds);
+
+    // set RGB value on mask of LEDs
+    void set_RGB(uint8_t chan, uint32_t ledmask, uint8_t red, uint8_t green, uint8_t blue);
+
+    /*
+      set a color on a 0 to 1 scale following a classic rainbow. Useful for scripting
+    */
+    void set_Rainbow(uint8_t chan, uint32_t ledmask, float v);
```

Comment:
This actually seems sorta clunky, why not leave this to be a scripting decision if it really wants it? (IE it's saddling everyone with the overhead of it, when it could be implemented in the script)

Reply:
because it will be very slow in a script I think. I think setting colours on rainbow is likely to be the most common use of LED scripting. Having to have the little interpolated lookup table in the script will complicate the script quite a lot

Reply:
Tested on CubeBlack running QuadPlane at 300Hz on the bench (and I checked that I at least got quasireasonable changes with SITL). Given the differences I don't think we need the custom binding. (I didn't remove the extra binding key when you look at the memory numbers, but it's actually quite notable that we can do the loop without allocating anymore memory then we otherwise would have.

Using the API binding:

```
APM: Lua: Running /APM/scripts/LED_roll.lua    
APM: Lua: Time: 3179 Mem: 29079 + 308    
APM: Lua: Running /APM/scripts/LED_roll.lua    
APM: Lua: Time: 1701 Mem: 29079 + 308    
APM: Lua: Running /APM/scripts/LED_roll.lua    
APM: Lua: Time: 3121 Mem: 29079 + 308    
APM: Lua: Running /APM/scripts/LED_roll.lua    
APM: Lua: Time: 2980 Mem: 29079 + 308    
APM: Lua: Running /APM/scripts/LED_roll.lua    
APM: Lua: Time: 3172 Mem: 29079 + 308    
APM: Lua: Running /APM/scripts/LED_roll.lua    
APM: Lua: Time: 3158 Mem: 29079 + 308    
APM: Lua: Running /APM/scripts/LED_roll.lua    
APM: Lua: Time: 3062 Mem: 29079 + 308    
APM: Lua: Running /APM/scripts/LED_roll.lua    
APM: Lua: Time: 3103 Mem: 29079 + 308    
APM: Lua: Running /APM/scripts/LED_roll.lua    
APM: Lua: Time: 3880 Mem: 29079 + 308    
APM: Lua: Running /APM/scripts/LED_roll.lua    
APM: Lua: Time: 3280 Mem: 29079 + 308    
APM: Lua: Running /APM/scripts/LED_roll.lua    
APM: Lua: Time: 2764 Mem: 29079 + 308    
APM: Lua: Running /APM/scripts/LED_roll.lua    
APM: Lua: Time: 3442 Mem: 29079 + 308    
APM: Lua: Running /APM/scripts/LED_roll.lua    
APM: Lua: Time: 3042 Mem: 29079 + 308    
APM: Lua: Running /APM/scripts/LED_roll.lua    
APM: Lua: Time: 2698 Mem: 29079 + 308    
APM: Lua: Running /APM/scripts/LED_roll.lua   
```

Using pure Lua:
```
APM: Lua: Time: 3531 Mem: 30603 + 308
APM: Lua: Running /APM/scripts/LED_roll.lua
APM: Lua: Time: 3306 Mem: 30603 + 308
APM: Lua: Running /APM/scripts/LED_roll.lua
APM: Lua: Time: 1680 Mem: 30603 + 308
APM: Lua: Running /APM/scripts/LED_roll.lua
APM: Lua: Time: 3157 Mem: 30603 + 308
APM: Lua: Running /APM/scripts/LED_roll.lua
APM: Lua: Time: 3089 Mem: 30603 + 308
APM: Lua: Running /APM/scripts/LED_roll.lua
APM: Lua: Time: 3325 Mem: 30603 + 308
APM: Lua: Running /APM/scripts/LED_roll.lua
APM: Lua: Time: 4571 Mem: 30603 + 308
APM: Lua: Running /APM/scripts/LED_roll.lua
APM: Lua: Time: 1730 Mem: 30603 + 308
APM: Lua: Running /APM/scripts/LED_roll.lua
APM: Lua: Time: 3776 Mem: 30603 + 308
APM: Lua: Running /APM/scripts/LED_roll.lua
APM: Lua: Time: 2971 Mem: 30603 + 308
APM: Lua: Running /APM/scripts/LED_roll.lua
APM: Lua: Time: 1780 Mem: 30603 + 308
APM: Lua: Running /APM/scripts/LED_roll.lua
APM: Lua: Time: 3273 Mem: 30603 + 308
APM: Lua: Running /APM/scripts/LED_roll.lua
APM: Lua: Time: 3142 Mem: 30603 + 308
APM: Lua: Running /APM/scripts/LED_roll.lua
APM: Lua: Time: 2950 Mem: 30603 + 308
APM: Lua: Running /APM/scripts/LED_roll.lua
APM: Lua: Time: 4027 Mem: 30603 + 308
APM: Lua: Running /APM/scripts/LED_roll.lua
APM: Lua: Time: 2960 Mem: 30603 + 308
```

The change to the script:

``` diff
diff --git a/libraries/AP_Scripting/examples/LED_roll.lua b/libraries/AP_Scripting/examples/LED_roll.lua
index b472ace5ca..05c0b98ede 100644
--- a/libraries/AP_Scripting/examples/LED_roll.lua
+++ b/libraries/AP_Scripting/examples/LED_roll.lua
@@ -34,6 +34,29 @@ function constrain(v, vmin, vmax)
    return v
 end
 
+local rainbow = {    
+  { 255, 0, 0 },
+  { 255, 127, 0 },
+  { 255, 255, 0 },
+  { 0,   255, 0 },
+  { 0,   0,   255 },
+  { 75,  0,   130 },
+  { 143, 0,   255 },
+}
+
+function set_Rainbow(chan, led, v)    
+  local num_rows = #rainbow
+  local row = constrain(v * num_rows, 1, num_rows)
+  local v0 = row / num_rows
+  local v1 = (row+1) / num_rows
+  local p = (v - v0) / (v1 - v0)
+  serialLED:set_RGB(chan,
+                    1 << led,
+                    (rainbow[row].rgb[0] + p * (float(rainbow[row+1].rgb[0]) - float(rainbow[row].rgb[0]))),
+                    (rainbow[row].rgb[1] + p * (float(rainbow[row+1].rgb[1]) - float(rainbow[row].rgb[1]))),
+                    (rainbow[row].rgb[2] + p * (float(rainbow[row+1].rgb[2]) - float(rainbow[row].rgb[2]))))
+end
+
 function update_LEDs()
 
   -- 16 LEDs per wing, two channels
@@ -43,8 +66,8 @@ function update_LEDs()
     local v_left  = constrain(0.5 * (1 - (roll/45) * (led / num_leds)), 0, 1)
     local v_right = constrain(0.5 * (1 + (roll/45) * (led / num_leds)), 0, 1)
 
-    serialLED:set_Rainbow(chan_left, 1<<led, v_left)
-    serialLED:set_Rainbow(chan_right,1<<led, v_right)
+    serialLED:set_Rainbow(chan_left, led, v_left)
+    serialLED:set_Rainbow(chan_right,led, v_right)
   end
   serialLED:send()
```